torch/utils/data/sampler.py

```diff
@@ -104,10 +104,14 @@ def num_samples(self):
 
     def __iter__(self):
         n = len(self.data_source)
+        if self.generator is None:
+            self.generator = torch.default_generator
```

Comment:
don't use the default generator. it is the same as using `None`. the problem of using the default generator is that the RNG consumption is nondeterministic since the #times this iterator is advanced depends on multiprocessing communication. instead, sample a seed from the current generator, create a new generator object with that seed. and don't overwrite the attr, just 
```py
if self.generator is None:
  generator = ...
else
  generator = self.generator
```